# SVM tutorial

## Libraries : Populating the iterative namespace

In [22]:
# Check required packages - Use pip install to install a new package 
!pip install numpy

In [23]:
from sklearn import svm, datasets
import numpy as np
import pylab as plt
%matplotlib inline

### Utils

## 1. Binary classification

In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt
import scipy.stats as st
import csv 

In [25]:
data = pd.read_csv('MoyennesFinales.csv',encoding='latin-1')

In [26]:
data.head()

Mm6  Mp6  Ms6  Mm5  Mp5  Ms5  Mm4  Mp4  Ms4  Mm3  Mp3  Ms3     MS  S
0   17   16   17   17   17   19   16   16   14   15   15   14  15.44  C
1   14   13   14   15   15   16   12   13   14   10    8    9  13.79  D
2   16   16   18   16   17   20   16   14   16   13   14   13  14.36  D
3   14   13   11   12   15   15   10   12   12   10   10    9  11.18  A
4   15   10   14   11   12   13   12   12   12    8    9    9  11.37  A

In [27]:
# créer la matrice de données
X = data.iloc[:,0:12].values

# créer le vecteur d'étiquettes
y = data.iloc[:,13].values 
y_class = np.where(y == 'A', 0, 1)

### Classifier instance 
Display default parameters

In [28]:
svc = svm.SVC(kernel='linear')  # svc = svm.LinearSVC()
print("Default parameters:")
svc

Default parameters:


SVC(kernel='linear')

### Paramaters tuning
* Kernel variation

In [29]:
X.shape

(175, 12)

In [32]:
#Nous découpons nos données en un jeu d'entrainement et un jeu de test

from sklearn import model_selection
X_train, X_test, y_train, y_test = \
    model_selection.train_test_split(X, y_class, test_size=0.2)

In [36]:
svc.kernel='linear'
svc.fit(X_train, y_train)

SVC(kernel='linear')

In [37]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.89


In [38]:
svc = svm.SVC(kernel='poly', gamma='scale', degree=3)
svc.fit(X_train, y_train)

SVC(kernel='poly')

In [39]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.83


In [40]:
svc = svm.SVC(kernel='rbf', gamma='scale') #, gamma=1)
svc.fit(X_train, y_train)

SVC()

In [41]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.80


Conclusion 
le linear est meilleur
pour la classification binaire avec 0,86

* Penalty parameter C of the error term

In [45]:
svc = svm.SVC(kernel='rbf', C=.01, gamma='scale')
svc.fit(X_train, y_train)

SVC(C=0.01)

In [46]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.63


In [47]:
svc = svm.SVC(kernel='rbf', C=100, gamma='scale')
svc.fit(X_train, y_train)

SVC(C=100)

In [48]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.83


In [49]:
svc = svm.SVC(kernel='linear', C=100, gamma='scale')
svc.fit(X_train, y_train)

SVC(C=100, kernel='linear')

In [50]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.86


In [53]:
svc = svm.SVC(kernel='poly', C=100, gamma='scale', degree=3)
svc.fit(X_train, y_train)

SVC(C=100, kernel='poly')

In [54]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.77


Conclusion?
C a amélioré rbf mais linear n'a pas changé et est toujours le meilleur

## 2. Multiclass classification

In [55]:
data = pd.read_csv('MoyennesBinaires.csv',encoding='latin-1')

In [56]:
data.head()

Mm6  Mp6  Ms6  Mm5  Mp5  Ms5  Mm4  Mp4  Ms4  Mm3  Mp3  Ms3  MS  S
0   17   16   17   17   17   19   16   16   14   15   15   14  15  2
1   14   13   14   15   15   16   12   13   14   10    8    9  14  1
2   16   16   18   16   17   20   16   14   16   13   14   13  14  1
3   14   13   11   12   15   15   10   12   12   10   10    9  11  0
4   15   10   14   11   12   13   12   12   12    8    9    9  11  0

In [57]:
# créer la matrice de données
X = data.iloc[:,0:12].values

# créer le vecteur d'étiquettes
y = data.iloc[:,13].values 

In [67]:
#Nous découpons nos données en un jeu d'entrainement et un jeu de test

from sklearn import model_selection
X_train, X_test, y_train, y_test = \
    model_selection.train_test_split(X, y, test_size=0.2)

In [68]:
svc = svm.SVC(kernel='linear')  # svc = svm.LinearSVC()
print("Default parameters:")
svc

Default parameters:


SVC(kernel='linear')

In [69]:
svc.kernel='linear'
svc.fit(X_train, y_train)

SVC(kernel='linear')

In [70]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.83


In [71]:
svc = svm.SVC(kernel='poly', gamma='scale', degree=3)
svc.fit(X_train, y_train)

SVC(kernel='poly')

In [72]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.77


In [73]:
svc = svm.SVC(kernel='rbf', gamma='scale') #, gamma=1)
svc.fit(X_train, y_train)

SVC()

In [74]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.77


Penalty parameter C of the error term

In [75]:
svc = svm.SVC(kernel='rbf', C=100, gamma='scale')
svc.fit(X_train, y_train)

SVC(C=100)

In [76]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.74


In [77]:
svc = svm.SVC(kernel='linear', C=100, gamma='scale')
svc.fit(X_train, y_train)

SVC(C=100, kernel='linear')

In [78]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.80


In [79]:
svc = svm.SVC(kernel='poly', C=100, gamma='scale', degree=3)
svc.fit(X_train, y_train)

SVC(C=100, kernel='poly')

In [80]:
from sklearn.metrics import accuracy_score

pred = svc.predict(X_test)
print("accuracy {:.2f}".format(accuracy_score(y_test, pred)))

accuracy 0.74


SVC multi-class works with "one versus one" strategy, k(k-1)/2 SVMs are build. LinearSVC uses "one versus all" strategy for multi-classification to run faster for such as sparse data sets, text mining is a typical example.

Conclusion?
linear sans penalty demeure le meilleur pour la classification multiclasse
0.83